pkg/store/tsdb.go

```diff
@@ -111,31 +114,46 @@ func (s *TSDBStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSer
 	return nil
 }
 
-func (s *TSDBStore) encodeChunk(it tsdb.SeriesIterator) (storepb.AggrChunk, error) {
-	chk := chunkenc.NewXORChunk()
+func (s *TSDBStore) encodeChunks(it tsdb.SeriesIterator, maxSamplesPerChunk int) (chks []storepb.AggrChunk, err error) {
+	var (
+		chkMint int64
+		chk     *chunkenc.XORChunk
+		app     chunkenc.Appender
+		isNext  = it.Next()
+	)
+
+	for isNext {
+		if chk == nil {
+			chk = chunkenc.NewXORChunk()
+			app, err = chk.Appender()
+			if err != nil {
+				return nil, err
+			}
+			chkMint, _ = it.At()
+		}
 
-	app, err := chk.Appender()
-	if err != nil {
-		return storepb.AggrChunk{}, err
-	}
-	var mint int64
+		app.Append(it.At())
+		chkMaxt, _ := it.At()
 
-	for i := 0; it.Next(); i++ {
-		if i == 0 {
-			mint, _ = it.At()
+		isNext = it.Next()
+		if isNext && chk.NumSamples() < maxSamplesPerChunk {
```

Comment:
Won't this be a problem if  there's a single chunk that doesn't reach maxSamplesPerChunk ?

Reply:
That's why "isNext" check here. We check if it reaches max AND there is next item. 

So if the iterator ends, we close the chunk even though it does not have max chunk items.